### PR TITLE
Feature/lifecycle docs

### PR DESCRIPTION
**Pull request by collijk**

### CODE REVIEW

File: src/vivarium/framework/lifecycle.py [hunk 1]
```diff
@@ -111,6 +187,7 @@ def __str__(self):
 
 
 class LifeCycle:
+    """A Concrete representation of the flow of simulation execution states."""
```
Comment (kfwilson) [inline src/vivarium/framework/lifecycle.py]: maybe replace the phrase on line 6 with this one?

File: src/vivarium/framework/lifecycle.py [hunk 2]
```diff
@@ -121,7 +198,25 @@ def __init__(self):
...
+            The list of names (in order) of the states that make up the
+            life cycle phase.  State names must be unique across the entire
+            life cycle.
+        loop
+            Whether the life cycle phase states loop.
+
+        Raises
+        ------
+        LifeCycleError
+            If the phase or state names are non-unique.
```
Comment (kfwilson) [inline src/vivarium/framework/lifecycle.py]: hmm so the docstring for LifeCycleError says `Error raised when lifecycle ordering contracts are violated.`  - is a duplicate state name an ordering violation?

File: src/vivarium/framework/lifecycle.py [hunk 3]
```diff
@@ -156,29 +286,57 @@ def _validate(self, phase_name: str, states: List[str]):
...
+        Constraints are applied by dynamically wrapping and binding a method
+        to an existing component at run time.
+
+        Parameters
+        ----------
+        method
+            The method to constrain.
+        permitted_states
+            The life cycle states in which the method can be called.
+
```
Comment (kfwilson) [inline src/vivarium/framework/lifecycle.py]: missing return documentation

File: src/vivarium/framework/lifecycle.py [hunk 4]
```diff
@@ -193,80 +351,233 @@ def _wrapped(*args, **kwargs):
...
+        Raises
+        ------
+        TypeError
+            If an unbound function is supplied for constraint.
+        ValueError
+            If the provided method is a python "special" method (i.e. a
+            method surrounded by double underscores).
+
```
Comment (kfwilson) [inline src/vivarium/framework/lifecycle.py]: missing return documentation

File: src/vivarium/framework/lifecycle.py [hunk 5]
```diff
@@ -193,80 +351,233 @@ def _wrapped(*args, **kwargs):
...
+            method is restricted from being called.
+
+        Raises
+        ------
+        ValueError
+            If neither ``allow_during`` nor ``restrict_during`` are provided,
+            or if both are provided.
+        LifeCycleError
+            If states provided as arguments are not in the life cycle.
+
```
Comment (kfwilson) [inline src/vivarium/framework/lifecycle.py]: add ConstraintError here

File: src/vivarium/framework/lifecycle.py [hunk 6]
```diff
@@ -193,80 +351,233 @@ def _wrapped(*args, **kwargs):
...
+    def __str__(self) -> str:
         return str(self.lifecycle)
 
 
 class LifeCycleInterface:
+    """Interface to the life cycle management system.
+
+    The life cycle management system allows components to constrain
+    methods so that they're only available during certain simulation
+    life cycle states.
```
Comment (kfwilson) [inline src/vivarium/framework/lifecycle.py]: maybe also a line about ordering here?

File: src/vivarium/framework/lifecycle.py [hunk 7]
```diff
@@ -193,80 +351,233 @@ def _wrapped(*args, **kwargs):
...
+            method is restricted from being called.
+
+        Raises
+        ------
+        ValueError
+            If neither ``allow_during`` nor ``restrict_during`` are provided,
+            or if both are provided.
+        LifeCycleError
+            If states provided as arguments are not in the life cycle.
+
```
Comment (kfwilson) [inline src/vivarium/framework/lifecycle.py]: add ConstraintError

File: src/vivarium/framework/lifecycle.py [hunk 8]
```diff
@@ -1,21 +1,55 @@
...
+from types import MethodType
+from typing import List, Callable, Tuple
 import textwrap
 
 from vivarium.exceptions import VivariumError
 
 
 class LifeCycleError(VivariumError):
-    """Error raised when lifecycle contracts are violated."""
+    """Error raised when life cycle ordering contracts are violated."""
```
Comment (kfwilson) [inline src/vivarium/framework/lifecycle.py]: It seems like this is also used when a life cycle phase / state is missing or duplicated which doesn't seem necessarily an ordering thing?

File: src/vivarium/framework/lifecycle.py (hunk 6 above)
Comment (collijk) [reply src/vivarium/framework/lifecycle.py]: Ordering contracts are handled privately in the simulation context.  If a user ever sees one, we're in trouble.

File: docs/source/concepts/lifecycle.rst [hunk 9]
```diff
@@ -4,98 +4,136 @@
...
+   * - | :ref:`Bootstrap <lifecycle_bootstrap>`
+     - | Arguments passed from the simulation
+       | :ref:`entry point <entry_points_concept>` are parsed and core
+       | framework systems are initialized.
+   * - | :ref:`Initialization <lifecycle_initialization>`
+     - | Simulation managers and :ref:`components <components_concept>` are
+       | initialized and all :ref:`configuration <configuration_concept>`
+       | information is loaded.
+   * - | :ref:`Setup <lifecycle_setup>`
+     - | Components request register themselves with simulation services and
```
Comment (halfhorst) [inline docs/source/concepts/lifecycle.rst]: extra word here

File: docs/source/concepts/lifecycle.rst [hunk 10]
```diff
@@ -4,98 +4,136 @@
...
-components have been instantiated.  This is a useful phase in the simulation
-lifecycle because you can typically modify what components are in the system
-or how they are configured without any consequences.
+
+.. _lifecycle_bootstrap:
+
+The Bootstrap Phase
+-------------------
+
+The bootstrap and initialization phase look like an atomic operation to an
```
Comment (halfhorst) [inline docs/source/concepts/lifecycle.rst]: either phase should be plural or look needs an 's'

File: docs/source/concepts/lifecycle.rst [hunk 11]
```diff
@@ -129,7 +172,7 @@ each :ref:`time step <time_concept>`:
    simulation outputs.
 
 By listening for these events, individual components can perform actions,
-including manipulating simulants. This sequence of events is repeated until
+including manipulating the. This sequence of events is repeated until
```
Comment (halfhorst) [inline docs/source/concepts/lifecycle.rst]: you dropped something, maybe simulants?

**Pull request merged**